pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,350 @@
+	if err != nil {
+		return err
+	}
+
+	err = b.checkForLockRace(stackRef)
+	if err != nil {
+		b.Unlock(stackRef)
+		return err
+	}
+
+	return nil
+}
+
+func (b *lockableBackend) Unlock(stackRef backend.StackReference) {
+	err := b.lb.bucket.Delete(context.TODO(), b.lockPath(stackRef.Name()))
+	if err != nil {
+		logging.Errorf("there was a problem deleting the lock at %v, things may have been left in a bad "+
+			"state and a manual clean up may be required: %v",
+			filepath.Join(b.lb.url, b.lockPath(stackRef.Name())), err)
+	}
+}
+
+func (b *lockableBackend) isLockForThisStack(file string, stackRef backend.StackReference) bool {
+	return strings.HasPrefix(file, b.lockPrefix(stackRef.Name()))
+}
+
+func (b *lockableBackend) lockDir() string {
+	return filepath.Join(workspace.BookkeepingDir, workspace.LockDir)
+}
+
+func (b *lockableBackend) lockPrefix(stack tokens.QName) string {
+	contract.Require(stack != "", "stack")
+	return filepath.Join(b.lockDir(), fsutil.QnamePath(stack)+".")
+}
+
+func (b *lockableBackend) lockPath(stack tokens.QName) string {
+	contract.Require(stack != "", "stack")
+	return b.lockPrefix(stack) + b.lockID + ".json"
+}
+
+func (b *lockableBackend) ListStacks(ctx context.Context,
+	projectFilter *tokens.PackageName) ([]backend.StackSummary, error) {
+	return b.lb.ListStacks(ctx, projectFilter)
+}
+func (b *lockableBackend) Query(ctx context.Context, stackRef backend.StackReference,
+	op backend.UpdateOperation) result.Result {
+	return b.lb.Query(ctx, stackRef, op)
+}
+
+func (b *lockableBackend) GetHistory(ctx context.Context,
+	stackRef backend.StackReference) ([]backend.UpdateInfo, error) {
+	return b.lb.GetHistory(ctx, stackRef)
+}
+
+func (b *lockableBackend) GetLogs(ctx context.Context, stackRef backend.StackReference,
+	cfg backend.StackConfiguration,
+	query operations.LogQuery) ([]operations.LogEntry, error) {
+	return b.lb.GetLogs(ctx, stackRef, cfg, query)
+}
+
+func (b *lockableBackend) GetLatestConfiguration(ctx context.Context,
+	stackRef backend.StackReference) (config.Map, error) {
+	return b.lb.GetLatestConfiguration(ctx, stackRef)
+}
+
+func (b *lockableBackend) GetStackTags(ctx context.Context,
+	stackRef backend.StackReference) (map[apitype.StackTagName]string, error) {
+	return b.lb.GetStackTags(ctx, stackRef)
+}
+func (b *lockableBackend) Logout() error {
+	return b.lb.Logout()
+}
+func (b *lockableBackend) CurrentUser() (string, error) {
+	return b.lb.CurrentUser()
+}
+
+func (b *lockableBackend) Name() string {
+	return b.lb.Name()
+}
+
+func (b *lockableBackend) URL() string {
+	return b.lb.URL()
+}
+
+func (b *lockableBackend) ParseStackReference(stackRefName string) (backend.StackReference, error) {
+	return b.lb.ParseStackReference(stackRefName)
+}
+
+func (b *lockableBackend) GetStack(ctx context.Context,
+	stackRef backend.StackReference) (backend.Stack, error) {
+	return b.lb.GetStack(ctx, stackRef)
+}
+
+func (b *lockableBackend) local() {
+}
```

Comment:
unused?

Reply:
required to implement the backend interface